main/src/ca/uwaterloo/flix/language/phase/PatternExhaustiveness.scala

```diff
@@ -0,0 +1,524 @@
+         *  }
+         *
+         * case Some True =>
+         * case Some False =>
+         * case None =>
+         *
+         * {Some, None} is a complete signature, but the "Some" case is exhaustive only if the {True, False} case is
+         *
+         * exhaustive. So we create a "Specialized" matrix for "Some" with {True, False} as rows and check that.
+         */
+        val checkAll: List[Exhaustiveness] = sigma.map(c => {
+          val res: Exhaustiveness = findNonMatchingPat(root, specialize(root, c, rules), countCtorArgs(c) + n - 1)
+          res match {
+            case a:Exhaustive => a
+            case NonExhaustive(ctors) => NonExhaustive(rebuildPattern(c, ctors))
+          }
+        })
+        checkAll.foldRight(Exhaustive(): Exhaustiveness)(mergeExhaustive)
+
+      } else {
+        /* If the constructors are not complete, then we will fall to the wild/default case. In that case, we need to
+         * check for non matching patterns in the wild/default matrix.
+         */
+        findNonMatchingPat(root, defaultMatrix(rules), n - 1) match {
+          case Exhaustive() => Exhaustive()
+          case NonExhaustive(ctors) => sigma match {
+            // If sigma is not empty, pick one of the missing constructors and return it
+            case _ :: _ => NonExhaustive(rebuildPattern(missing.head, ctors))
+            // Else, prepend a wildcard
+            case Nil => NonExhaustive(rebuildPattern(Wild(), ctors))
+          }
+        }
+      }
+    }
+
+    /**
+      * Specialize a matrix of patterns for the Constructor ctor For a constructor of C(r1,...ra) and a matrix of width
+      * n, we return a matrix of width n+a-1
+      *
+      * e.g. If we have
+      * enum Option {
+      *    case Some a,
+      *    case Nothing
+      *  }
+      *
+      * And the pattern matrix
+      *
+      * case Some True =>
+      * case Some False =>
+      * case Nothing =>
+      *
+      * Specializing for Some gives use the matrix
+      * True =>
+      * False =>
+      *
+      * Where' we've taken a matrix had made of the arguments to the Some rows. Specializing for Nothing would give
+      * an empty matrix as it has no arguments.
+      *
+      * @param root The ast root
+      * @param ctor The constructor to specialize for
+      * @param rules The rules matrix to specialize
+      * @return The specialized matrix of rules
+      */
+    def specialize(root: TypedAst.Root, ctor: TypeConstructor, rules: List[List[TypedAst.Pattern]]): List[List[TypedAst.Pattern]] = {
+      // First figure out how many arguments are needed by the ctor
+      val numArgs = countCtorArgs(ctor)
+
+      def specializeRow(pat: List[TypedAst.Pattern], acc: List[List[TypedAst.Pattern]]) = pat.head match {
+        // A wild constructor is the same as the constructor
+        // with all its arguments as wild
+        case a: TypedAst.Pattern.Wild =>
+          (List.fill(numArgs)(a) ++ pat.tail) :: acc
```

Comment:
We prefer to use `:::` for list append since `++` is overloaded and can cause issues with type inference when you start changing the types around.